### PR TITLE
move css, fonts, js to assets folder; auto install assets folder

### PR DESCRIPTION
**Pull request by Clonkk**
PR for #12 with auto deploy of assets folder

### CODE REVIEW

**Comment by Clonkk** [comment]
@pietroppeter 
This is a draft because while working on it, I noticed some weird stuff when having 2 nimibook outside of the repo.

When having two books ``booka`` and ``bookb`` I had reference on ``booka`` when working with ``bookb``.
I think some path used in Nimibook gets generated compile time and do not get re-compiled when swapping book, giving you weird side effects (I haven't fully identified yet).

File: nimibook.nimble [hunk 1]
```diff
@@ -14,27 +14,32 @@ requires "nimib >= 0.1.2"
...
 import os
+
 task genbook, "build book":
   selfExec(" r -d:release genbook.nim")
 
 task dumpbook, "dump book.json":
   selfExec(" r -d:release -d:dumpBook genbook.nim")
 
 task cleanbook, "remove all files created during build":
+  # Since assets now gets populated, I think we can just delete everything in docs
```
Comment (HugoGranstrom) [inline nimibook.nimble:25]: Me and @pietroppeter discussed this at the bottom of #12 that we perhaps shouldn't delete the entire `docs/` folder but only the relevant files. Otherwise, it's hard to incorporate static (non-Nimib generated) files such as static images or stylesheets into the book. We'd have to create some way of adding user-assets to the `docs/` folder in that case like you did for the built-in assets if we are going the `rm docs` path.

**Comment by Clonkk** [comment]
Fair enough

**Comment by HugoGranstrom** [comment]
It would make sense though to have an explicit `assets` folder in the future though, which is copied into `docs/` on generation of the book 😄 And then we could remove it all

**Comment by Clonkk** [comment]
I kept the assets folder I just didn't delete docs but docs/assets instead

**Comment by HugoGranstrom** [comment]
Okay, that should work for a while 😄

**Comment by Clonkk** [comment]
Other solution would be to have custom static file in a dedicated folder like  ``static`` remove everything except this.

On another topic, when testing with a dummy book outside of the repo, all the html file are created in the current folder and not n ``docs``. What am I missing for the html to be generated inside ``docs`` ?

**Comment by pietroppeter** [comment]
a few quick notes

* I would keep the `init` step which creates/updates the initial structure separate from the `build` step. After init you commit the assets and they only change if you want to update nimibook
* the nimble file should be modified using `installDirs` to keep some folders during installation. Ideally I would create a single `init` directory that has the initial empty structure with all files, folders and assets needed. init should be called with a "book name" that will use it to initialize some fields
* ideally I would add in this init folder also the `config.nims` and gitignore and the rest, basically everything except for stuff contained in book folder (and we should move templates to a template folder and change accordingly `nbDoc.templateDirs`. The files in `init` might probably need to be templates themselves (so that you would substitute book name in apprpriate places)
* probably the missing ingredients for wrong folder output is a nimble file. nimble file is what nimib uses to anchor what the project folder is. even an empty nimble file would work

**Comment by Clonkk** [comment]
> I would keep the init step which creates/updates the initial structure separate from the build step. After init you commit the assets and they only change if you want to update nimibook

For automated setup from scratch, you will need the explicit init therefore the init will have to be comitted so automated deployment from ``git clone`` can work. That's why I think we should init everytime.

What we can do instead is populate ``assets`` only and only if it does not exists.
It's what I do with Mustache template that are only copied from if they do not exists.

> the nimble file should be modified using installDirs to keep some folders during installation. Ideally I would create a single init directory that has the initial empty structure with all files, folders and assets needed. init should be called with a "book name" that will use it to initialize some fields

``installDirs``with ``srcDir="src`` will cause Nimble to complain on installation (because of https://github.com/nim-lang/nimble/issues/653).

So either
*  we stop using ``srcDir`` and flatten the repo (but then nimble will complain about ``genbook`` being top level...) 
* We add the folders in ``src/nimibook/supername`` and we don't need ``installDirs`` -> I vote for this option personnally. 


> ideally I would add in this init folder also the config.nims and gitignore and the rest, basically everything except for stuff contained in book folder (and we should move templates to a template folder and change accordingly nbDoc.templateDirs. The files in init might probably need to be templates themselves (so that you would substitute book name in apprpriate places)

I can regroup them in a  ``init`` (or  ``ressources`` ?) folders ``init/assets`` and ``initmustache``).

> probably the missing ingredients for wrong folder output is a nimble file. nimble file is what nimib uses to anchor what the project folder is. even an empty nimble file would work

I don't like the dependence on Nimble. Not all books hould be nimble package IMO. How about using : 
```nim
template newBookFromToc*(booklabel: string, rootfolder: string, body: untyped): Book =
  initBook(rootfolder)
  static:
    let path_to_rootfolder = getProjectPath() / rootfolder
    putEnv("nimibook_path_to_rootfolder", path_to_rootfolder)
```
and then : 

```nim
proc useNimibook*(nbDoc: var NbDoc) =
  # path handling (fix upstream in nimib)
  let
    nbThisFile = changeFileExt(nbDoc.filename.AbsoluteFile, ".nim")
    thisTuple = nbThisFile.splitFile
    pathToRootFolder = getEnv("nimibook_path_to_rootfolder")
    nbThisDir: AbsoluteDir = pathToRootFolder.toAbsoluteDir
    nbHomeDir: AbsoluteDir = nbThisDir / RelativeDir("..") / "docs".RelativeDir
    nbSrcDir = nbThisDir

```
Also, since we now have explicit path (obtained during the compilation of the user book), we can probably simplify some stuff (because  we won't need relativeTo anymore)

**Comment by Clonkk** [comment]
Unless I misunderstood somethingI think there's an issue with the init specialization.
Even after removing reference to ``findNimbleDir`` in ``nimibook`` (and replacing it with a correct path) it still break without ``nimble`` file present.

**Comment by pietroppeter** [comment]
> > I would keep the init step which creates/updates the initial structure separate from the build step. After init you commit the assets and they only change if you want to update nimibook
> 
> For automated setup from scratch, you will need the explicit init therefore the init will have to be comitted so automated deployment from `git clone` can work. That's why I think we should init everytime.

Automated setup from scratch you mean like in our CI? I might be understanding wrong our CI but I think what is doing is rebasing on main before generating the documentation (the build step). If we commited in main the result of an "update" step (the init but just to update to more recent assets/templates...), I think the CI for gh-pages should be able to incorporate that and do not need an init step.

What I am thinking of is the possible user workflows and their relative frequencies:

* the user will very frequently want to regenerate a single page (already appearing in the toc). This is currently possible by just running `nim r path/to/file.nim` (before my last PR this required a separate creation of toc/book.json)
* the user will rather frequently want to regenerate the whole book (every time it changes the toc or some book configuration). this is currently possible with `nimble genbook`
* only occasionally, the user will need to update assets.

> What we can do instead is populate `assets` only and only if it does not exists.
> It's what I do with Mustache template that are only copied from if they do not exists.

this would be needed if we decide to go with init every time.

> 
> > the nimble file should be modified using installDirs to keep some folders during installation. Ideally I would create a single init directory that has the initial empty structure with all files, folders and assets needed. init should be called with a "book name" that will use it to initialize some fields
> 
> `installDirs`with `srcDir="src` will cause Nimble to complain on installation (because of [nim-lang/nimble#653](https://github.com/nim-lang/nimble/issues/653)).
> 
> So either
> 
> * we stop using `srcDir` and flatten the repo (but then nimble will complain about `genbook` being top level...)
> * We add the folders in `src/nimibook/supername` and we don't need `installDirs` -> I vote for this option personnally.

I did know `installDirs` could have issues and I look foward to the `frictionless` package structure.
I agree with your pick on putting the asset stuff inside a subfolder of `src` (mdbook does the same btw, there is a `theme` folder with templates and assets)

> 
> > ideally I would add in this init folder also the config.nims and gitignore and the rest, basically everything except for stuff contained in book folder (and we should move templates to a template folder and change accordingly nbDoc.templateDirs. The files in init might probably need to be templates themselves (so that you would substitute book name in apprpriate places)
> 
> I can regroup them in a `init` (or `ressources` ?) folders `init/assets` and `init/mustache`).

yes, that would be ok. I would use `init/templates` instead of `init/mustache`
for `init` I am not sure if this is the best name. we could follow the lead of `mdbook` and use `theme` also (they do not split templates from assets but I think it is better).

> 
> > probably the missing ingredients for wrong folder output is a nimble file. nimble file is what nimib uses to anchor what the project folder is. even an empty nimble file would work
> 
> I don't like the dependence on Nimble. Not all books hould be nimble package IMO. How about using :
> 
> ```nim
> template newBookFromToc*(booklabel: string, rootfolder: string, body: untyped): Book =
>   initBook(rootfolder)
>   static:
>     let path_to_rootfolder = getProjectPath() / rootfolder
>     putEnv("nimibook_path_to_rootfolder", path_to_rootfolder)
> ```
> 
> and then :
> 
> ```nim
> proc useNimibook*(nbDoc: var NbDoc) =
>   # path handling (fix upstream in nimib)
>   let
>     nbThisFile = changeFileExt(nbDoc.filename.AbsoluteFile, ".nim")
>     thisTuple = nbThisFile.splitFile
>     pathToRootFolder = getEnv("nimibook_path_to_rootfolder")
>     nbThisDir: AbsoluteDir = pathToRootFolder.toAbsoluteDir
>     nbHomeDir: AbsoluteDir = nbThisDir / RelativeDir("..") / "docs".RelativeDir
>     nbSrcDir = nbThisDir
> ```
> 
> Also, since we now have explicit path (obtained during the compilation of the user book), we can probably simplify some stuff (because we won't need relativeTo anymore)

yeah, using the nimble file mechanism is also one of those things I want to rediscuss in nimib (possibly a wrong design decision that we can still reverse).

Something like you have there it is a fine workaround for now (if we could avoid an env variable it would be better but I am not sure how). Note though that `nbThisDir` is the directory of the current processed nim file (e.g. for `basics\plotting.nim` is `basics` and we should not mix it with `nbSrcDir`).

In summary, agree on most points, only pushing back a little on the `init` always thing. :)

**Comment by HugoGranstrom** [comment]
I'm not a fan of using ENV variables either, feels like some obscure bugs could creep up in the future. Couldn't we just add a field in `book.json` with the path? The only downside I see is that it has to be regenerated if moved between devices (as CI) but the CI should run it either way.

**Comment by Clonkk** [comment]
> Automated setup from scratch you mean like in our CI? I might be understanding wrong our CI but I think what is doing is rebasing on main before generating the documentation (the build step). If we commited in main the result of an "update" step (the init but just to update to more recent assets/templates...), I think the CI for gh-pages should be able to incorporate that and do not need an init step.

No, I didn't meant like our CI :). Our CI is based on on github actions and can commit to a branch.  

If I want to use Nimibook on a self-hosted server and automate deployment, I need to be able to generate everything from a zip/tarball/clean git clone; using only Bash (whatever scripting language your OS uses).

Conditionnal init that is required but only once make this harder - the workaround would be to manually implement it in each genbook file with ``when defined(...)`` .

> What I am thinking of is the possible user workflows and their relative frequencies:

>   *  the user will very frequently want to regenerate a single page (already appearing in the toc). This is currently possible by just running nim r path/to/file.nim (before my last PR this required a separate creation of toc/book.json)
    
 >  * the user will rather frequently want to regenerate the whole book (every time it changes the toc or some book configuration). this is currently possible with nimble genbook
    
  > * only occasionally, the user will need to update assets.


If ``assets`` are deployed automatically from templated versions then it can only change if you change Nimibook versions or customize it.

So it shouldn't really matter if we re-populate assets (beside maybe the 0.2 milliseconds it takes to copy the folders) in the init everytime with the condition that the init gets skipped for files that already exists (because customization).

Realistically, I think what will end up happening is that we will write this in ``genbook.nim`` : 
```nim
if initCond:
  initBook
``` 
To avoid having to modify source code depending on where we are in the deployment process. 

So why not incorporate it already in the library ? I really don't see the downsides of it.

> I'm not a fan of using ENV variables either, feels like some obscure bugs could creep up in the future. Couldn't we just add a field in book.json with the path? The only downside I see is that it has to be regenerated if moved between devices (as CI) but the CI should run it either way.

What's the issue with env variables ? It's not persistent after the program ends and multiple program running in multiple shell concurrently will have their own environment (you can check it with multiple ``inim`` running concurrently - they do not modify each other's environment).

**Comment by HugoGranstrom** [comment]
> What's the issue with env variables ? It's not persistent after the program ends and multiple program running in multiple shell concurrently will have their own environment (you can check it with multiple inim running concurrently - they do not modify each other's environment).

My bad then, didn't do my research on what it meant in this context. I thought it was environment variables as in `path` and `JAVA_HOME` 🙃 The only situation this wouldn't work is if we generate and publish the toc in different files, but that shouldn't happen :)

File: src/nimibook/docs.nim [hunk 2]
```diff
@@ -7,11 +7,18 @@ proc useNimibook*(nbDoc: var NbDoc) =
...
     nbThisFile = changeFileExt(nbDoc.filename.AbsoluteFile, ".nim")
     thisTuple = nbThisFile.splitFile
-    nbThisDir: AbsoluteDir = thisTuple.dir
-    nbHomeDir: AbsoluteDir = findNimbleDir(nbThisDir) / "docs".RelativeDir
-    nbSrcDir = nbHomeDir / RelativeDir(".." / "book")
-
-  # are these two actually needed? well, home_path is needed in path_to_root, but other than that?
+    pathToRootFolder = getEnv("nimibook_path_to_rootfolder")
+    nbThisDir: AbsoluteDir = pathToRootFolder.toAbsoluteDir
+    nbHomeDir: AbsoluteDir = nbThisDir / RelativeDir("..") / "docs".RelativeDir
```
Comment (HugoGranstrom) [inline src/nimibook/docs.nim:12]: `nbHomeDir` is never used, could that be the problem?

File: src/nimibook/entries.nim [hunk 3]
```diff
@@ -1,4 +1,4 @@
-import std / [os, strutils, strformat]
```
Comment (HugoGranstrom) [inline src/nimibook/entries.nim:1]: This breaks it on Windows as `string.replace` is in `strutils`

**Review by HugoGranstrom — CHANGES_REQUESTED** [review]

File: src/nimibook/docs.nim (hunk 2 above)
Comment (HugoGranstrom) [reply src/nimibook/docs.nim:12]: [This](https://github.com/pietroppeter/nimib/blob/dda83a58e1fb2ae53523e29905614f9bfa62a02a/src/nimib.nim#L77) line uses the `nbHomeDir` injected by `nbInit` to decide where to write the file.

File: src/nimibook/entries.nim (hunk 3 above)
Comment (Clonkk) [reply src/nimibook/entries.nim:1]: Oopsie

**Comment by pietroppeter** [comment]
> No, I didn't meant like our CI :). Our CI is based on on github actions and can commit to a branch.
> 
> If I want to use Nimibook on a self-hosted server and automate deployment, I need to be able to generate everything from a zip/tarball/clean git clone; using only Bash (whatever scripting language your OS uses).

ok, then I did not understand. I created a new RFC #20 for the design of a nimibook cli. With that design, to generate the book starting from scratch from a git clone of an existing repository  would require the following commands:

- `nim c nbook`
- `nbook init`
- `nbook build`

> What's the issue with env variables ? It's not persistent after the program ends and multiple program running in multiple shell concurrently will have their own environment (you can check it with multiple inim running concurrently - they do not modify each other's environment).

You are right, env variables like that are good. As @HugoGranstrom was mentioning I think putting a path in `book.json` would also be ok, but I would avoid putting an aboslute path in case. The advantage of using `book.json` with current structure would be to be able to still run `nim r book/basics/plotting.nim` and the file is generated (if `plotting.nim` is not called from `genbook` it will not use any env variable). With the target structure proposed in #20 there is no particular difference (you would call `nbook build basics/plotting.nim`). Still since we have already a `book.json` that is a channel to communicate from main driver to single chapter, I think we could keep using that channel and not use anotehr mechanism.

Note also that in #20 I propose to drop current mechanism in `genbook` based on compile time switches (which requires to recompile) and propose to pass to a real cli.

File: src/nimibook/docs.nim (hunk 2 above)
Comment (Clonkk) [reply src/nimibook/docs.nim:12]: Good catch. So it's something to be fixed in Nimib.

**Comment by Clonkk** [comment]
To sum-up for this PR : 
* [ ] Fix Nimib using injected nbHome in ``nbSave``
* [ ] Init that does nothing if file structure is already existing. 
* [ ] Path (avoiding absolute path) in book.json
* [ ] Make a CLI to avoid recompiling (optionnal maybe it'd be best to tackle the issues separately. I can give it a go and we'll see).

**Comment by pietroppeter** [comment]
I was thinking maybe the fix in nimib could indirectly come through [this refactoring](https://github.com/pietroppeter/nimib/issues/25)? If you have a global `nb` object you have an easy way to change everything (you will be able to change `nb.HomeDir` and that it will be used). In particular `useNimibook` in `nimibook / docs` would change the signature to use the `nb` object instead of `nbDoc` and the changes that you make in there for `nbHomeDir` would then be seen outside.

To be honest, I have not been able to understand the bug reported here, so let me just state what I know:
* in nimib `nbHomeDir` is initialized during `nbInit` to be either the folder with the closes parent nimble file or the `docs` folder inside of it (or the folder containing the source if no nimble file is found)
* then `nbHomeDir` is set to current directory (so that if you write/read a file with a relative path it will be relative to that)
* on top of that, as @HugoGranstrom noticed, I also use `with(nbHomeDir)` to make sure that current directory is still `nbHomeDir` before saving (and saving happens with `nbDoc.filename` as path which is usually a relative path).
* the lines 8-12 in current `nimibook / docs` are just a workaround for the missing `nbSrcDir` in nimib. the definition of `nbHomeDir` [here](https://github.com/pietroppeter/nimib/issues/25) is not though as to be injected but just as a more "straightforward" (ehm...) mechanism to define `nbSrcDir` (previous lines 8-10 are basically to reconstruct `nbHomeDir` from `nbDoc.filename` knowing what the assumptions are on how it was created during `nbInit`).

mmh, okay after stating all this I might actually understand the bug. This is due to the missing nimble file in the initialized empty structure, right?

**Comment by HugoGranstrom** [comment]
> mmh, okay after stating all this I might actually understand the bug. This is due to the missing nimble file in the initialized empty structure, right?

Yes indeed. Nimib sets `nbHomeDir = findNimble() (/ "docs")` by default and then writes the files relative to `nbHomeDir`. So unless we change `nbHomeDir` on the NimiBook side, it will just be the current directory if no `.nimble` file is found. 

So yes a global `nb` object would solve this as it allows us to intercept and change `nbHomeDir` in a proc :) (We could still change it in the `nbUseNimibook` template though)

**Comment by Clonkk** [comment]
> You are right, env variables like that are good. As @HugoGranstrom was mentioning I think putting a path in book.json would also be ok

Do I override ``book.path_to_root `` or do you want to add a new field ? ?

**Comment by Clonkk** [comment]
Okay this is mostly done I think.

There were a few changes for the CLI mode, so this PR needs to be reviewed carefully (@pietroppeter).

I'm just waiting on the Nimib thing to get resolved to remove the placeholder nimble files in examples.

**Comment by HugoGranstrom** [comment]
Nice work! :D

Would this PR to Nimib suffice in the meantime we wait for the global nb object? https://github.com/pietroppeter/nimib/pull/53
With it we can pass in the absolute path to nbHomeDir and nbSrcDir in the compile flags.

**Comment by Clonkk** [comment]
Yeah I think it should suffice

File: book/.gitignore [hunk 4]
```diff
@@ -1,2 +1,3 @@
 *.html
+*.mustache
```
Comment (pietroppeter) [inline book/.gitignore:2]: why ignore the mustache files?

File: nbook.nim [hunk 5]
```diff
@@ -13,4 +13,5 @@ var book = newBookFromToc("nimibook", "book"):
     draft("Draft chapter", "draft.md")
 
 book.git_repository_url = "https://github.com/pietroppeter/nimibook"
-nbBookTasks
\ No newline at end of file
+book.keep = @["tocexample"]
```
Comment (pietroppeter) [inline nbook.nim:16]: do we need this, since it appears in the toc?

File: src/nimibook/books.nim [hunk 6]
```diff
@@ -51,18 +51,90 @@ proc dumpHook*(s: var string, v: string) =
     s.setLen(at)
 
 proc dump*(book: Book) =
+  var book = book
+  let rootFolder = getEnv("nimibook_rootfolder")
```
Comment (pietroppeter) [inline src/nimibook/books.nim:55]: in the end, didn't we decide to use `book.json` to pass the path?

File: src/nimibook/docs.nim [hunk 7]
```diff
@@ -20,16 +41,13 @@ proc useNimibook*(nbDoc: var NbDoc) =
   # debugEcho "Current directory: ", getCurrentDir()
   # debugEcho "Output file: ", nbDoc.filename
 
-
   # templates are in nbSrcDir
   nbDoc.templateDirs = @[nbSrcDir.string]
```
Comment (pietroppeter) [inline src/nimibook/docs.nim:45]: templates are staying in book folder? I thought we were moving them to a templates but maybe that was an assumption I made in my head and never shared... :)

**Review by pietroppeter — COMMENTED** [review]
thanks a lot for this! I added some questions.

Other remarks:

* on general folder structure, I was thinking, maybe we could put in src/nimibook a `theme` folder which contains a full example structure of a default nimibook (with folders `docs\assets`, `templates`, `book`, ...). If we add to that a clever config.nims (with a path `../../` we probably can also run nbook there and test the default structure during development). The theme folder is then the reference from where we populate assets.
* on nimib file, on second thoughts I think it could still make sense to keep it as a requirement. Yes, not every book should be a package, but when built with nimib, it will have dependencies and those naturally belong in a nimble file. Also tasks in a nimble file can be useful (but that I guess could be done with nims file, so this probably does not count). I honestly like better the nimble file used as anchor for root than other alternatives.

Some of the questions or above remarks maybe can be addressed in a further PR, the goal of this if I am not wrong is to be able to get a mechanism to initialize and update a nimibook project. If otherimprovements are left for a next PR, then we test the update mechanism...

File: nbook.nim (hunk 5 above)
Comment (Clonkk) [reply nbook.nim:16]: No, we don't it was just to check that the ``book.keep`` mechanism was working

File: book/.gitignore (hunk 4 above)
Comment (Clonkk) [reply book/.gitignore:2]: Mustache file gets copied from ``ressources/template`` so I hardly see the point of comitting them in the book folder

File: src/nimibook/docs.nim (hunk 7 above)
Comment (Clonkk) [reply src/nimibook/docs.nim:45]: Maybe we misunderstood each other. My idea was to have a ``ressources/template`` folder in ``src/nimibook`` to deploy the mustache files automatically in the book folder (per the init mechanism we discussed)

File: src/nimibook/books.nim (hunk 6 above)
Comment (Clonkk) [reply src/nimibook/books.nim:55]: This i what I use to write to the path in the json. ``book.path_to_root = getEnv("...")`` and then we write book as Json.

Comment (Clonkk) [reply src/nimibook/books.nim:55]: The only issues is ``useNimibook`` also uses rootFolder in the same process (other process have the value passed by --define) and needs to know the rootfolder

Comment (pietroppeter) [reply src/nimibook/books.nim]: ah ok. as in the [comment above](https://github.com/pietroppeter/nimibook/pull/19#pullrequestreview-674953517) (not in the review). I would actually still keep the nimble file mechanism. can we revert this part?

File: src/nimibook/docs.nim (hunk 7 above)
Comment (pietroppeter) [reply src/nimibook/docs.nim:38]: in [this comment](https://github.com/pietroppeter/nimibook/pull/19#pullrequestreview-674953517) I describe an alternative folder structure. Do you think it could be done in this PR? if not we can do it/discuss it later. If it stays like this maybe I would just change `ressources` to `resources` :)

**Comment by Clonkk** [comment]
> on nimib file, on second thoughts I think it could still make sense to keep it as a requirement. Yes, not every book should be a package, but when built with nimib, it will have dependencies and those naturally belong in a nimble file. 

Not everybody use nimble as a dependency manager. Status dropped his usage internally, other people made their own such as : 
* https://github.com/disruptek/nimph
* https://github.com/Araq/nawabs
There are others, this isn't an exhaustive lists.

* Also, it's not intuitive. I have fairly good understanding of Nimib and Nimibook, and still couldn't figure out why it wasn't working. I assume less informed users are going to be even more lost.



> Also tasks in a nimble file can be useful (but that I guess could be done with nims file, so this probably does not count). 

Tasks and CLI mode serve the exact same purpose. We have a CLI based application; we don't need to impose Tasks anymore (which doesn't mean they can't be used).



> I honestly like better the nimble file used as anchor for root than other alternatives.

While Tasks can be useful the whole point of making a CLI based module is to do what Tasks 



> I honestly like better the nimble file used as anchor for root than other alternatives.


I really dislike it as a solution for Nimibook.

1) It forces users to use Nimble for documentation. It will deter everyone who dislike / don't need / don't want to use Nimble.
Many users (me included), uses Nimble only as a distribution mechanism (i.e. only uses ``nimble install``, ``nimble develop``) because working with Nimble on bigger projects is seriously a PITA. On bigger projects, I now rely on a [dist](https://github.com/disruptek/dist)-like system based on submodule and ``--path`` in cfg files.  


2) It's unreliable because edge-case / non-convetionnal directory structure will break code for no reason.
Typically, how would you handle mono-repo multi-package with no top level Nimble file or nested Nimble package ?


3) It's unnecessary. Working under the assumption that the binary run is applied to the current folder is perfectly reasonable for a CLI (Most Linux tools works this way : ``grep``, ``find``, ``ls``, ``rm`` etc.). 


We only assumes  ``getCurrentFolder()`` is the base folders of your book and ``nbook`` is ran from there - which is what you do anyway with Nimble because you can't run Nimble command from another folder.

Using ``getCurrentDir()`` is way more intutive for a book structure.


4) It doesn't follow the KISS principle. Keep It Simple Stupid - those are the best APIs, no shenanigans, no weird edge-case, no convoluted stuff for hypothetical use-case that never happens in practice.
 
Even hard-coding ``docs`` and ``books`` (removing the needs for all of this) for Nimibook would be acceptable (and better than ``findNimbleDir``) because realistically Nimibook users will just copy the structure from this repo anyway.

File: src/nimibook/docs.nim (hunk 7 above)
Comment (Clonkk) [reply src/nimibook/docs.nim:38]: > on general folder structure, I was thinking, maybe we could put in src/nimibook a theme folder which contains a full example structure of a default nimibook (with folders docs\assets, templates, book, ...). If we add to that a clever config.nims (with a path ../../ we probably can also run nbook there and test the default structure during development). The theme folder is then the reference from where we populate assets.

If I understand correctly, you want to have the whole book structure in ``src/nimibook/theme`` and copy it in the current folder upon calling the ``init`` / ``update`` step ?

That is perfectly do-able, I don't think we even need a config.nims to test it (Init should do the trick).

**Comment by pietroppeter** [comment]
thanks for the honest and thorough feedback. I will take this into account also when reviewing path handling in nimib.

For nimibook, it is ok to proceed with current PR as it is. I think the missing part was the override mechanism for nbHomeDir (https://github.com/pietroppeter/nimib/pull/53) which has been merged and released in nimib v0.1.3.

**Comment by HugoGranstrom** [comment]
In the README I think it's worth noting that if you update the TOC you will have to recompile the CLI. It's obvious if you think about how it works but people don't always think 😛

**Comment by Clonkk** [comment]
Good point ! I'll add it.

**Comment by Clonkk** [comment]
CI is green, the github page deployed on my fork looks good and everything works locally in my machine !

I think we're good to go :).

File: README.md [hunk 8]
```diff
@@ -20,22 +20,33 @@ provided by [nim-markdown]. For a quick reference of supported syntax see the [c
 
 ### Status
 
-nimibook currently provides minimal functionality to create a book, notably missing is
-a command line interface to automate common task (e.g. initialization of a new nimibook).
+nimibook currently provides minimal functionality to create a book and support a local CLI mode.
 
 ## Installation
 
 To be able to use nimibook, currently the only way is to recreate the same content of the repository
```
Comment (pietroppeter) [inline README.md:27]: I think we can update the installation instruction, right? it should be: 1) install nimibook 2) create a simple nim file (`nbook.nim`) with content (`import nimibook; nbBookTasks`) and run `nim r nbook init` in the folder where you want the book. Should we also rename `nbBookTasks` to `nimibookCli`?

File: src/nimibook/docs.nim (hunk 7 above)
Comment (pietroppeter) [reply src/nimibook/docs.nim:37]: let's do and discuss this in another PR :)

File: README.md (hunk 8 above)
Comment (Clonkk) [reply README.md:27]: Seems reasonable. I'll modify it after lunch :)

File: src/nimibook/resources/templates/document.mustache [hunk 9]
```diff
@@ -58,7 +58,7 @@
     <body>
         <!-- Provide site root to javascript -->
         <script type="text/javascript">
-            var path_to_root = "{{ path_to_root }}";
+            var path_to_root = "{{ path_to_root }}/assets";
```
Comment (pietroppeter) [inline src/nimibook/resources/templates/document.mustache:61]: I guess whether or not this is correct depends on how the other js scripts (which I have not looked at) use and interpret the path_to_root? if it is only to interact with assets this is correct. If it is only to interact with the pages, the original is not correct. If they use it for both uses, we might have to reconsider the use of assets folder inside docs (or patch the js scripts)

Comment (Clonkk) [reply src/nimibook/resources/templates/document.mustache:61]: A quick grep in the resources folder shows that ``path_to_root`` is only used ``document.mustache``  so the js scripts should not use ``path_to_root``, I guess ?

To the best of my understanding (which is not great - javascript is not my friend), ``path_to_root`` is only used to call java script in the generated html.

**Comment by Clonkk** [comment]
@pietroppeter  I just realized nimibookCli uses ``”ook`` explicitly so how would you feel aboud ``newBookFromToc`` injecting the ``book`` variable instead of relying on ``var book = newBookFromToc()`` ?

Basically, this
```nim
# var book = ... # Use another variable name instead
var my_super_book = newBookFromToc(...):

nimibookCli
```
will break the CLI.

Alternatively, we can pass book as a paramters to ``nimibookCli(book)``.

**Comment by HugoGranstrom** [comment]
IMO `nimibookCli(book)` feels more direct and then the user will get a compile error if they forget it

**Comment by Clonkk** [comment]
Either solution is fine IMO.
One will look like this : 

```nim
newBookFromToc(...):
  ...
  
book.git_repository_url = "https://github.com/pietroppeter/nimibook"
nimibookCli
```

The other will be : 
```nim
var mybook = newBookFromToc(...):
  ...
  
mybook.git_repository_url = "https://github.com/pietroppeter/nimibook"
nimibookCli(mybook)
```

**Comment by pietroppeter** [comment]
> Either solution is fine IMO.
>
> I committed solution n°2 for now (to check CI only fails because of that).

agree both solution are fine. 2nd is more explicit and ok to proceed with that.

File: src/nimibook/resources/templates/document.mustache (hunk 9 above)
Comment (pietroppeter) [reply src/nimibook/resources/templates/document.mustache:61]: yeah, it seems a false alarm. ok for me.

**Comment by pietroppeter** [comment]
I think we are good to go, right? should I merge? 🥁

File: README.md [hunk 10]
```diff
@@ -20,22 +20,34 @@ provided by [nim-markdown]. For a quick reference of supported syntax see the [c
...
+nimibook currently provides minimal functionality to create a book and support a local CLI mode.
 
 ## Installation
 
-To be able to use nimibook, currently the only way is to recreate the same content of the repository
-and adjust it to your content. You can do that using the template feature from github
-or with a checkout of the repo and copying and pasting.
-
-Discussion of further options is ongoing in [issue #12](https://github.com/pietroppeter/nimibook/issues/12).
+To install Nimibook simply use : ``nimble install nimibook``
```
Comment (pietroppeter) [inline README.md:27]: ah wait, we need to tell the user how to write the initial `nbook.nim` that will allow him to init everything!

**Comment by pietroppeter** [comment]
thanks @Clonkk for the work on this and the patience.

# 🚀 🥳
